### ISSUE TITLE
Matrix Solve Behavior w/ Row of 0's in Augmented Matrix

### ISSUE THREAD

**Issue by Auptimum** [markrogoyski/math-php#413]
Hello! As the title implies, I've encountered a problem with the Matrix->solve() method when the augmented matrix contains a row of 0's. I'm not sure if this is possible to remedy because it might just be the behavior of the lib. The augmented matrix of A and b is below:

[1,         0,          0,          0,          0,          0,        |  1.457]
[0,         1,          0,          0,          1,          0,        | -1.457]
[0,         0,          1,          0,          0,          1,        | -5.664]
[0,         0,          0,          0,          0,          0,        |          0]
[0,         0,-180.92,          0,          0,-854.14,        | 1620.7]
[0,180.92,           0,          0, 854.14,          0,         | -416.8]

You can see the 4th row (and column) is entirely 0. Below is my expected output after calling solve():

[ 1.457]
[-1.230]
[-4.779]
[    0    ]
[-0.227]
[-0.885]

However, the solver method gives the below output:
[ 1.457]
[-1.230]
[-4.779]
[-0.227]
[-0.885]
[    0    ]

I can see why this is the case as the determinant of the matrix is 0, but some online solvers give a solution better for my usage, such as this one: https://matrix.reshish.com/gaussSolution.php (image below)
![image](https://user-images.githubusercontent.com/56129386/110648416-b2c83c80-8186-11eb-8af8-9f4d9ce9b7b2.png)

I know I can get a proper solution by reducing the matrix, eliminating the 4th row and whichever column is a zero vector (it could be column 1 or 4) to get to a square matrix with a non-zero determinant. After calling solve, I can then fill in the 0's myself to get back to a 6x1 solution vector. Was just wondering about this; if it is bad behavior to have in other situations. I'm no linear algebra veteran.

Thanks for the contribution to the PHP community!

**Comment by markrogoyski** [comment]
Hi @Auptimum,

Thanks for your interest in MathPHP.

I was able to reproduce your matrix solve result. All the numbers are right, but the order seems to be an issue. If we use the result of the solve, x, and plug it into Ax = b, we don't get b using the order with the 0 at the end of the vector. However, if the 0 is in the zero-based third position of the Vector, then the result solves Ax = b.

This is a singular matrix, and the preferred LU/QR factorization methods don't work. So the library resorts to doing elementary row operations to reduce the matrix into REF and then RREF form to solve the system of linear equations. Row reduction operations are problematic when dealing with floating point numbers. Anyways, I think it's just taking the right most column and calling that x, which is usually the correct thing to do, but in this case, it doesn't line up with the variables because of the row full of zeros. RREF ends up looking like this:
```php
[1, 0, 0, 0, 0, 0, 1.457]
[0, 1, 0, 0, 0, 0, -1.2294375984077]
[0, 0, 1, 0, 0, 0, -4.7787483437806]
[0, 0, 0, 0, 1, 0, -0.22756240159235]
[0, 0, 0, 0, 0, 1, -0.88525165621936]
[0, 0, 0, 0, 0, 0, 0]
```
I think what we need to do is reformat the RREF so the ones are along the diagonal and then take the rightmost column and that will be the correct solution x in this case.

I'll investigate this a bit and see if this makes sense and works for other interesting matrices that have to resort to row reduction.

For your information, other popular libraries just throw errors on this matrix because it is singular. For example:

R
```R
> A
     [,1]   [,2]    [,3] [,4]   [,5]    [,6]
[1,]    1   0.00    0.00    0   0.00    0.00
[2,]    0   1.00    0.00    0   1.00    0.00
[3,]    0   0.00    1.00    0   0.00    1.00
[4,]    0   0.00    0.00    0   0.00    0.00
[5,]    0   0.00 -180.92    0   0.00 -854.14
[6,]    0 180.92    0.00    0 854.14    0.00
> b
[1]    1.457   -1.457   -5.664    0.000 1620.700 -416.800
> solve(A, b)
Error in solve.default(A, b) : 
  Lapack routine dgesv: system is exactly singular: U[4,4] = 0
```

Python Scipy
```python
>>> A
[[1, 0, 0, 0, 0, 0], [0, 1, 0, 0, 1, 0], [0, 0, 1, 0, 0, 1], [0, 0, 0, 0, 0, 0], [0, 0, -180.92, 0, 0, -854.14], [0, 180.92, 0, 0, 854.14, 0]]
>>> b
[1.457, -1.457, -5.664, 0, 1620.7, -416.8]
>>>
>>> scipy.linalg.solve(A, b)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.9/site-packages/scipy/linalg/basic.py", line 214, in solve
    _solve_check(n, info)
  File "/usr/local/lib/python3.9/site-packages/scipy/linalg/basic.py", line 29, in _solve_check
    raise LinAlgError('Matrix is singular.')
numpy.linalg.LinAlgError: Matrix is singular.
```

Python Numpy
```python
>>> A
[[1, 0, 0, 0, 0, 0], [0, 1, 0, 0, 1, 0], [0, 0, 1, 0, 0, 1], [0, 0, 0, 0, 0, 0], [0, 0, -180.92, 0, 0, -854.14], [0, 180.92, 0, 0, 854.14, 0]]
>>> b
[1.457, -1.457, -5.664, 0, 1620.7, -416.8]
>>> np.linalg.solve(A, b)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<__array_function__ internals>", line 5, in solve
  File "/usr/local/lib/python3.9/site-packages/numpy/linalg/linalg.py", line 393, in solve
    r = gufunc(a, b, signature=signature, extobj=extobj)
  File "/usr/local/lib/python3.9/site-packages/numpy/linalg/linalg.py", line 88, in _raise_linalgerror_singular
    raise LinAlgError("Singular matrix")
numpy.linalg.LinAlgError: Singular matrix
```

These other libraries just throw an exception. Singular matrices are problematic for solving. MathPHP's matrix solve tries many strategies to solve a matrix, starting with LU and QR and finally resorting to elementary row operations to RREF nothing else is working. There might be other strategies to try before RREF like least squares approximation or some pseudo inverse via SVD. But for now RREF is our last resort, and I'll see what I can do to make it work even on this problematic singular matrix. I'll update this thread when I have a solution that works.

Mark

**Comment by Auptimum** [comment]
Hey, Mark! Thanks for the reply! And for making PHPMath more intelligent than Num/Scipy; I like idiot-proofness.

I think an alternative to messing with the RREF solver is maybe to use the column position of the 1's in the RREF to determine the position in the solution vector, instead of the right-most column's row positions. In this case, there's a 1 at columns 1, 2, 3, 5, and 6, which correspond to the positions for my desired solution vector.  I think this is appropriate as I understand the rows in the RREF for a system of equations to represent the corresponding coefficients for each x in each equation. Not 100% sure on that, but it is convenient in this case. 

Moreover, using the column positions of the 1's would also work in the case where the solution is defined, as the column positions would still correctly map the right-hand column to the solution vector.

The other, less elegant way I can think of is checking for "useless" row & column pairs, or just useless rows. As I mention in my case, both row 4 and column 4 are completely 0. Eliminating them before solving would actually make the RREF into a 5x5 square matrix where the 1's are along the diagonal, as they should be in the ideal case.

However, I think it's also enough to just get rid of rows that are zero-vectors. With row 4 gone, the RREF of the augmented matrix becomes:
```
[1, 0, 0, 0, 0, 0, 1.457]
[0, 1, 0, 0, 0, 0, -1.2294375984077]
[0, 0, 1, 0, 0, 0, -4.7787483437806]
[0, 0, 0, 0, 1, 0, -0.22756240159235]
[0, 0, 0, 0, 0, 1, -0.88525165621936]
```
So it still gets the right results, but we know that element 4 of b was removed. Using that knowledge, the answers can be mapped back to the resultant x vector, placing a 0 in the corresponding element of x (or ```INF``` - whatever makes more sense). However, I think this is a naive approach as it's coming from my less-experienced perspective, it seems artificial, and I prefer my 1st suggestion. I only offered it since it removes the "undefined" row and just wanted to brainstorm a bit.

Thanks for your time and thoughts, though!

**Comment by Auptimum** [comment]
Hey @markrogoyski, I came back to think about this. I refreshed myself on the subject recently, and the wikipedia [article ](https://en.wikipedia.org/wiki/Augmented_matrix#Solution_of_a_linear_system) on solving a system of equations via an augmented matrix has made me more confident that the RREF columns correspond to the solution vector positions. I think it's best to take advantage of this fact to get the proper x-vector out of the RREF solver. Have you had a chance to think about this some more? Does this sound reasonable?

**Comment by markrogoyski** [comment]
Hi @Auptimum,

I think it makes sense and I will implement something like that if the matrix is singular. I'll try to get an update out soon. I'll keep you posted in this thread when it is available in a new release. Sorry for the delay.

Mark

**Comment by markrogoyski** [comment]
Hi @Auptimum,

This has been fixed in [v1.11.0](https://github.com/markrogoyski/math-php/releases/tag/v1.11.0) and in the newest release [v2.0.0](https://github.com/markrogoyski/math-php/releases/tag/v2.0.0).

Thank you for reporting this issue. And thanks again for your interest in MathPHP.

Mark

**Comment by Auptimum** [comment]
Thanks, Mark! Appreciate it!
